text/0000-fun-vs-fun-ptr.md

```diff
@@ -0,0 +1,130 @@
+before 1.0. Note, while it is tempting to use DST for this, but unfortunately that would make
+function pointers fat, and moreover the sizes of functions are not in-general known dynamically
+anyway (c.f. foreign functions).
+
+So if this requires a breaking change to make nice, yet is too big to do before 1.0, what can be
+done? My suggestion is simply to make just the breaking changes now, and leave the rest for later:
+change the syntax of function pointer types to make clear they act like `&'statics`, and make it so
+that only "borrowed functions" coerce to function pointers. This doesn't force any semantic changes
+in the future, but opens to door to reusing `fn(...)...` for unboxed functions by allowing function
+pointers to silently become actual `&statics` if/when function types are added in the
+future. Meanwhile Rust, while slightly less ergonomic---function pointers are quite rare, is more
+forthright about its current expressiveness today.
+
+# Detailed design
+
+1. Change the syntax of function pointer types from
+
+   ```rust
+   ('extern' STRING_LIT?)? 'fn' '(' (IDEN ':' TYPE (',' IDEN ':' TYPE)* ','?)? ')' ('->' TYPE)?
+   ```
+
+   to
+
+   ```rust
+   '&' '\'static' ('extern' STRING_LIT?)? 'fn' '(' (IDEN ':' TYPE (',' IDEN ':' TYPE)* ','?)? ')' ('->' TYPE)?
+   ```
+
+   Like slices before DST, the `fn(...) ...` itself will not denote a type on its own and need not
+   even parse.
+
+   In the future, `fn(...) ...` can be made legal syntax and given semantics, in which case the
+   current syntax would no longer be its own grammatical production but remain valid, and actually
+   denote the type of a static, immutable borrowed pointer to a function.
+
+
+2. Change the coercion rules so that values of the secret function types to not coerce to function
+   pointers, but values of immutably borrowed secret function types do instead.
+
+   Just as today, these secret function types won't have a surface syntax. This change however
+   brings the coercion rules for functions in line with that for DSTs, and will prevent what would
+   be a much more auto-referencing if the types of functions were exposed in the
+   future. [Thanks @P1start for this idea.]
+   
+## Examples
+
+```rust
+fn f() { }
+// f : an undenotable, (ideally) zero sized, type. Think of it as a function "proxy".
+// &'static fn() is a function pointer type.
+// fn() could *someday* be an unsized function type.
+fn main() {
+    let g = f; // DEPENDS; should the proxy impls Copy?
+    let g = &f; // OK; g : &(an undenotable, (ideally) zero sized, type)
+    let g: &'static fn() = f; // ERROR: no auto-borrowing in Rust
+    let g: &'static fn() = &f; // OK; pointers to concrete function proxy coerce to pointers function
+    let g: &Fn() = &f; // OK; &magic -coerce-> &'static fn() -coerce-> &Fn trait object. Would necessitate dyanmic dispatch though.
+    let g: &Fn() = f; // DEPENDS; Should proxies themselves Fn?
```

Comment:
As @CloudiDust pointd out, this doesn't work, e.g. `let g: &Fn() = || {}` is not valid: it has to be `let g: &Fn() = &|| {};`, even though the closure `|| {}` implements `Fn`.